src/test/java/org/mockito/internal/stubbing/defaultanswers/ReturnsSmartNullsTest.java

```diff
@@ -85,4 +90,36 @@ public void should_print_the_parameters_on_SmartNullPointerException_message() t
                 .hasMessageContaining("lumpa");
         }
     }
+
+    // https://github.com/mockito/mockito/issues/357
+    @Test
+    public void generic_return_type_erasure() throws Throwable {
+        class S<T> {
+            public T f() {
+                return null;
+            }
+        }
+
+        S<Boolean> s = Mockito.mock(S.class, org.mockito.Answers.RETURNS_SMART_NULLS);
+
+        Mockito.when(s.f()).thenReturn(true);
+        assertTrue(s.f());
+    }
+
+    // https://github.com/mockito/mockito/issues/357
+    @Test
+    public void inferred_generic_type() throws Throwable {
+        class F {
+            <T> T g(Iterable<T> xs) {
```

Comment:
I think this is the way to go. Maybe we should even warn when mocking methods with a generic type argument when using smart nulls.